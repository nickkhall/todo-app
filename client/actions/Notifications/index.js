import { guidGenerator } from 'utils/formatter';

import {
  NOTIFICATION_ADD,
  NOTIFICATION_REMOVE
} from '../types';

/**
 * Toggles a notification to the user.
 * @param {boolean} notification Whether to show or hide notification.
 * @param {string} message Custom message to show user.
 * @param {string} type The type of notification to display.
 * @param {number=} duration The optional amount of of time show the message.
 */
export const toggleNotification = (message, type, duration) => (dispatch) => {
  const id = guidGenerator();

  dispatch({
    type: NOTIFICATION_ADD,
    payload: { message, type, id }
  });

  setTimeout(() => {
    dispatch({
      type: NOTIFICATION_REMOVE,
      payload: id
    });
  }, duration || 5000);
};
