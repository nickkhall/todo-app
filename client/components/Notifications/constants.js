export const DEFAULT_STACK_LIMIT = 3;
export const BACKGROUND_OPACITY = 0.8;
