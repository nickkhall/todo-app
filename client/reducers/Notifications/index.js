import {
  NOTIFICATION_ADD,
  NOTIFICATION_REMOVE
} from '../../actions/types';

export const initialState = {
  notifications: []
};

const addNotification = (state, payload) => {
  const { notifications } = state;

  return {
    ...state,
    notifications: notifications.concat(payload)
  };
};

const removeNotification = (state, id) => {
  const { notifications } = state;

  return ({
    ...state,
    notifications: notifications.filter(n => n.id !== id)
  });
};

const reducer = (state = initialState, action) => {
  const { payload, type } = action;

  switch (type) {
    case NOTIFICATION_ADD:
      return addNotification(state, payload);
    case NOTIFICATION_REMOVE:
      return removeNotification(state, payload);
    default:
      return state;
  }
};

export default reducer;
