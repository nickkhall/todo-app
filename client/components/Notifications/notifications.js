import React from 'react';

// MUI Components
import Fade from '@material-ui/core/Fade';

// Constants
import { DEFAULT_STACK_LIMIT } from './constants';

const Notification = ({
  classes,
  notifications,
  stackLimit = DEFAULT_STACK_LIMIT
}) => {
  if (!notifications || !notifications.length) return null;

  // Limit the notification stack to the user-defined limit, or the default
  // stack limit.
  const stack = notifications.slice(
    Math.max(notifications.length - stackLimit, 0)
  );

  return (
    <div className={classes.wrapper}>
      {
        stack.map(({ message, type, id }) => (
          <Fade
            key={id}
            in
            timeout={{ enter: 1000, exit: 1000 }}
          >
            <div className={classes.notification}>
              <div className={`${classes.content} ${classes[type]}`}>
                {message}
              </div>
            </div>
          </Fade>
        ))
      }
    </div>
  );
};

export default Notification;
