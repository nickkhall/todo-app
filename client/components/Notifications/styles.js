// MUI Colors
import amber from '@material-ui/core/colors/amber';
import blue from '@material-ui/core/colors/blue';
import green from '@material-ui/core/colors/green';
import red from '@material-ui/core/colors/red';

// Utils
import { hexToRGB } from 'utils/styles';

// Constants
import { BACKGROUND_OPACITY } from './constants';

export default theme => ({
  wrapper: {
    display: 'flex',
    flexDirection: 'column',
    pointerEvents: 'none', /* Allow click to bleed through to elements below */
    position: 'absolute',
    top: '10px',
    zIndex: 4000,

    /* On larger displays, notifications appear at the bottom left. */
    [theme.breakpoints.up('sm')]: {
      top: 'auto',
      bottom: '10px',
      left: 0,
      minWidth: '400px'
    }
  },
  content: {
    color: '#fff',
    fontSize: '1.1em',
    padding: '20px',
    margin: '10px',
    borderRadius: '3px',
    pointerEvents: 'auto'
  },
  success: {
    backgroundColor: hexToRGB(green[600], BACKGROUND_OPACITY)
  },
  error: {
    backgroundColor: hexToRGB(red[600], BACKGROUND_OPACITY)
  },
  info: {
    backgroundColor: hexToRGB(blue[600], BACKGROUND_OPACITY)
  },
  warning: {
    backgroundColor: hexToRGB(amber[700], BACKGROUND_OPACITY)
  }
});
