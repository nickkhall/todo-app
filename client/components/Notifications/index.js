
// MUI Components
import { withStyles } from '@material-ui/core/styles';

// Local styles
import styles from './styles';

import Notifications from './notifications';

export default withStyles(styles)(Notifications);
