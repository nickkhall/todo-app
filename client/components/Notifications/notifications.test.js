// MUI Components
import Fade from '@material-ui/core/Fade';

// Test utils
import {
  getComponentWrapper,
  testCommonComponentAttrs
} from 'utils/tests';

// Components
import Notifications from './notifications';

const defaultProps = { classes: {} };

const mockStackLimit = 2;

const generateMessage = () => ({
  message: Date.now(),
  type: 'error',
  id: Date.now()
});

let wrapper;
let props;

describe('<Notifications /> component', () => {
  testCommonComponentAttrs(Notifications, defaultProps);

  describe('basic rendering', () => {
    beforeEach(() => {
      props = {
        ...defaultProps
      };

      wrapper = getComponentWrapper(Notifications, props);
    });

    it('should render the proper amount of notifications', () => {
      props.notifications = [
        generateMessage(),
        generateMessage(),
        generateMessage()
      ];

      wrapper = getComponentWrapper(Notifications, props);
      expect(wrapper.find(Fade)).toHaveLength(props.notifications.length);
    });

    it('should limit notifications to the stack limit', () => {
      props.notifications = [
        generateMessage(),
        generateMessage(),
        generateMessage()
      ];

      props.stackLimit = mockStackLimit;

      wrapper = getComponentWrapper(Notifications, props);
      expect(wrapper.find(Fade)).toHaveLength(mockStackLimit);
      expect(wrapper.find(Fade)).not.toHaveLength(props.notifications);
    });

    it('should return empty without any notifications', () => {
      wrapper = getComponentWrapper(Notifications, props);
      expect(wrapper.find(Fade)).toHaveLength(0);
    });
  });
});
