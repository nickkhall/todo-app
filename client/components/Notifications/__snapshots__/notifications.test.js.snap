// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notifications /> component when rendering should render a component matching the snapshot 1`] = `""`;
